tensorizer/serialization.py

```diff
@@ -0,0 +1,1071 @@
+            device: Union[str, torch.device, None] = None,
+            dtype: Optional[str] = None,
+            filter_func: Optional[Callable[[str], Union[bool, Any]]] = None,
+    ) -> int:
+        """
+        Given `m`, a torch.nn.Module, load the associate tensors in this
+        Tensorizer object into the `torch.nn.Module`. Returns the number of
+        tensors loaded into the module.
+
+        Args:
+            m: The module to load the tensors into.
+            device: The device to load the tensors onto.
+            dtype: The dtype to load the tensors as. Defaults to None, which
+                means the dtype is not changed from the serialized dtype.
+            filter_func: A function (tensor_name: str) -> bool that returns
+                True if a tensor should be loaded, or False if it should be
+                skipped.
+        """
+        if device is None:
+            device = self._device
+            if device is None:
+                device = utils.get_device()
+
+        modules: typing.OrderedDict[str, torch.nn.Module] = OrderedDict()
+
+        for name, module in m.named_modules():
+            modules[name] = module
+
+        tensor_ct = 0
+
+        for name in self.keys():
+            if filter_func is not None and not filter_func(name):
+                continue
+            obj_path, attr = name.rsplit(".", 1)
+            module: torch.nn.Module = modules[obj_path]
+            entry = self._metadata[name]
+            param = self._to_torch_parameter(self.get(name), dtype, device)
+            if entry["type"] is TensorType.PARAM:
+                module.register_parameter(attr, param)
+            elif entry["type"] is TensorType.BUFFER:
+                module.register_buffer(attr, param)
+            elif entry["type"] is TensorType.STATE_DICT:
+                raise NotImplementedError(
+                    "This was serialized using the write_state_dict() method, and"
+                    " cannot be loaded using the load_tensors() method. Use the"
+                    " state_dict() method instead.")
+            tensor_ct += 1
+
+        self._file.close()
+        return tensor_ct
+
+
+class TensorSerializer:
+    """
+    Given a file-like object or path, serialize tensors from a torch.nn.Module
+    to it.
+
+    Args:
+        file_obj: A file-like object or path to a file to write to. The path
+            can be a S3 URI.
+        compress_tensors: If True, compress the tensors using lz4. This
+            exists as an internal curiousity as it doesn't seem to make
+            much of a difference in practice.
+    """
+
+    def __init__(
+            self,
+            file_obj: Union[io.BufferedIOBase, io.RawIOBase, typing.BinaryIO,
+                            str, bytes, os.PathLike, int],
+            compress_tensors: bool = False) -> None:
+        if isinstance(file_obj, (str, bytes, os.PathLike, int)):
+            self._file = stream_io.open_stream(file_obj, "wb+")
```

Comment:
Having the default mode be `wb+` will cause `open_stream` to throw an exception when using an S3 path, which is what happens in our serialization example.

We should either make the default mode `wb` or `ab` for everything or have separate logic to check if the `file_obj` is an `S3` path.

Reply:
This is fixed in my branch that will have a PR soon. S3 URIs weren't supposed to reject the `+` mode (in fact, it is required for calculating hashes currently, so it even adds a `+` if it's missing).